apis/python/src/tiledbsoma/dense_nd_array.py

```diff
@@ -148,16 +151,64 @@ def read_tensor(
         """
         Read a user-defined dense slice of the array and return as an Arrow ``Tensor``.
         """
-        tiledb_result_order = tiledb_result_order_from_soma_result_order(
-            result_order, accept=["column-major", "row-major"]
-        )
         with self._tiledb_open("r") as A:
             target_shape = _dense_index_to_shape(coords, A.shape, result_order)
-            query = A.query(return_arrow=True, order=tiledb_result_order)
-            arrow_tbl = query.df[coords]
-            return pa.Tensor.from_numpy(
-                arrow_tbl.column("soma_data").to_numpy().reshape(target_shape)
-            )
+            schema = A.schema
+
+        sr = clib.SOMAReader(
+            self._uri,
+            name=self.__class__.__name__,
+            result_order=result_order,
+            platform_config={} if self._ctx is None else self._ctx.config().dict(),
+        )
+
+        if coords is not None:
+            if not isinstance(coords, (list, tuple)):
+                raise TypeError(
+                    f"coords type {type(coords)} unsupported; expected list or tuple"
+                )
+            if len(coords) < 1 or len(coords) > schema.domain.ndim:
+                raise ValueError(
+                    f"coords {coords} must have length between 1 and ndim ({schema.domain.ndim}); got {len(coords)}"
+                )
+
+            for i, coord in enumerate(coords):
+                #                # Example: coords = [None, 3, slice(4,5)]
+                #                # coor takes on values None, 3, and slice(4,5) in this loop body.
+                dim_name = schema.domain.dim(i).name
+                if coord is None:
+                    pass  # No constraint; select all in this dimension
+                elif isinstance(coord, int):
+                    sr.set_dim_points(dim_name, [coord])
+                elif isinstance(coord, slice):
+                    lo_hi = util.slice_to_range(coord)
+                    if lo_hi is not None:
+                        lo, hi = lo_hi
+                        if lo < 0 or hi < 0:
+                            raise ValueError(
+                                f"slice start and stop may not be negative; got ({lo}, {hi})"
+                            )
+                        if lo > hi:
+                            raise ValueError(
+                                f"slice start must be <= slice stop; got ({lo}, {hi})"
+                            )
+                        sr.set_dim_ranges(dim_name, [lo_hi])
+                    # Else, no constraint in this slot. This is `slice(None)` which is like
+                    # Python indexing syntax `[:]`.
+                elif isinstance(
+                    coord, (collections.abc.Sequence, pa.Array, pa.ChunkedArray)
```

Comment:
Apologies in advance if I'm confused - but we changed the DenseNdArray to _only_ allow sub-array read/write, and not allow point indexing. In other words, passing in a pa.Array for coordinates is bogus for a dense array.  Point indexing is only supported for SparseNdArray and DataFrame.

This code (lies 198-201) will never be executed, as an exception will be raised in _dense_index_to_shape() if the user submits coordinates as an array.

Assuming my recollection (sub-array only) is correct, then there appear to be several bugs here (one is left over from my own PR - blush):
1. the `DenseNdCoordinates` type should only allow int and slice, and should not include `pa.Array`.
2. for consistency, the type signature of `_dense_index_to_shape()` should type the `coords` param with `DenseNdCoordinates`
3. the code supporting pa.Array (lines 198-201) should be removed.
4. And, a unit test should confirm all this (lacking one, we failed to notice these issues)